### PR TITLE
fix(destroy): Properties updated in a beforeDestroy hook are now pers…

### LINKED ISSUE #9318: destroy() in paranoid mode does not persist properties updated in a beforeDestroy hook
I'm opening a pull request for this in a minute, but I wanted to raise the issue first.

## What are you doing?

I have a model in paranoid mode. This model gets a `beforeDestroy` hook attached in order to update a property on deletion (`updatedBy` which is a FK to another model, but this is not related to the problem described here).
When calling `destroy()`, instances are correctly soft deleted by my changed property is not persisted.

```js
const ParanoidUser = this.sequelize.define('ParanoidUser', {
  username: Sequelize.STRING,
  updatedBy: Sequelize.INTEGER
}, {
  paranoid: true,
  hooks: {
    beforeDestroy: instance => instance.updatedBy = 1
  }
});

return ParanoidUser.sync({ force: true })
  .then(() => ParanoidUser.create({ username: 'user1' }))
  .then(user => user.destroy())
  .then(() => ParanoidUser.findOne({ paranoid: false }))
  .then(user => console.log(user.updatedBy));
```

## What do you expect to happen?
_1_

## What is actually happening?
_null_

__Dialect:__ mysql / sqlite
__Dialect version:__ 5.7
__Database version:__ 5.7
__Sequelize version:__ 4.36.1
__Tested with latest release:__ Yes, 5.0.0-beta2

**Note :** _Your issue may be ignored OR closed by maintainers if it's not tested against latest version OR does not follow issue template._

### PR DESCRIPTION
**Pull request by ddolcimascolo**
…isted on soft delete. Closes #9318

### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](../CONTRIBUTING.md)?

### Description of change

When getting the list of things to update in the database to perform a soft delete, the code is now considering changed values (through `Model.changed()`) instead of only updating `deletedAt`.

Closes issue #9318

### CODE REVIEW

**Comment by ddolcimascolo** [comment]
I think this should be applied on `4.x` branch also, but I'm unsure if I should submit a new PR targetting another branch, or if you automagically cherry-pick commits to backport...

Just tell me, I'm happy to open another PR if needs be.

File: test/integration/model.test.js [hunk 1]
```diff
@@ -1433,6 +1433,26 @@ describe(Support.getTestDialectTeaser('Model'), () => {
       });
     });
 
+    it('sets other changed values when soft deleting and a beforeDestroy hooks kicks in', function() {
```
[the remark this answers is not in the document]
Comment (ddolcimascolo) [reply test/integration/model.test.js]: Done

File: lib/model.js [hunk 2]
```diff
@@ -3896,7 +3896,7 @@ class Model {
       if (this.constructor._timestampAttributes.deletedAt && options.force === false) {
         const attribute = this.constructor.rawAttributes[this.constructor._timestampAttributes.deletedAt];
         const field = attribute.field || this.constructor._timestampAttributes.deletedAt;
-        const values = {};
+        const values = Utils.mapValueFieldNames(this, this.changed() || [], this.constructor);
```
[the remark this answers is not in the document]
Comment (ddolcimascolo) [reply lib/model.js]: Done

File: test/integration/model.test.js [hunk 3]
```diff
@@ -1433,6 +1433,26 @@ describe(Support.getTestDialectTeaser('Model'), () => {
...
+      const ParanoidUser = this.sequelize.define('ParanoidUser', {
```
[the remark this answers is not in the document]
Comment (ddolcimascolo) [reply test/integration/model.test.js]: Done

File: test/integration/hooks/destroy.test.js [hunk 4]
```diff
@@ -76,6 +76,51 @@ describe(Support.getTestDialectTeaser('Hooks'), () => {
         });
       });
     });
+
+    describe('with paranoid mode enabled', () => {
+
```
Comment (sushantdhiman) [inline test/integration/hooks/destroy.test.js]: remove extra line

File: test/integration/hooks/destroy.test.js [hunk 5]
```diff
@@ -76,6 +76,51 @@ describe(Support.getTestDialectTeaser('Hooks'), () => {
...
+        return this.ParanoidUser.sync({ force: true })
+          .then(() => this.ParanoidUser.create({ username: 'user1' }))
+          .then(user => user.destroy())
+          .then(() => this.ParanoidUser.findOne({ paranoid: false }))
+          .then(user => {
+            expect(user.virtualField).to.equal(0);
+          });
+      });
+
```
Comment (sushantdhiman) [inline test/integration/hooks/destroy.test.js]: remove extra line

File: test/integration/hooks/destroy.test.js [hunk 6]
```diff
@@ -76,6 +76,51 @@ describe(Support.getTestDialectTeaser('Hooks'), () => {
...
+    });
+
```
Comment (sushantdhiman) [inline test/integration/hooks/destroy.test.js]: remove extra line

**Comment by ddolcimascolo** [comment]
@sushantdhiman Done, sorry for the extra blank lines

**Comment by ddolcimascolo** [comment]
Thanks for the appoval, what about the backport to v4 ?

**Comment by sushantdhiman** [comment]
> Thanks for the appoval, what about the backport to v4 ?

I'll cherry pick commit from master and port to v4 :)

**Comment by sushantdhiman** [comment]
I was checking other use of this hook, we need to apply this change to https://github.com/sequelize/sequelize/blob/4caa17c78bbbe3212e65aaf6e58ea9aa1bcb10b5/lib/model.js#L2546-L2568

So `individualHooks: true` cases can use this as well. Will need extra tests.

Also `Model.destroy` is using is `Utils.now` but `Model.prototype.destroy` is setting `new Date`, probably both need to use `Utils.now`

This change may be unexpected for some users and definitely change hook behaviour (https://github.com/sequelize/sequelize/issues/9318#issuecomment-382788201) so its better to keep this in v5

**Comment by ddolcimascolo** [comment]
Fine to keep this v5 only.
Regarding [your latest comments](https://github.com/sequelize/sequelize/pull/9319#issuecomment-383029902) I can handle the changes and add tests for the bulk destroy use cases. Will try to do this later this day and update the PR.

Cheers

**Comment by ddolcimascolo** [comment]
@sushantdhiman Sorry, I'm puzzled :grinning: 

`Model.destroy` will indeed call the `beforeDestroy` hook on every instance impacted by the bulk destroy, but then how is the code supposed to "know" that some values changed on the instances, how is the code supposed to translate that in the bulk UPDATE statement that is issued to the database ?
We could even imagine that the hook has changed different attributes on different instances...

I'm thinking that the code should individually call `destroy` on each instance instead of doing a bulk operation, if and only if `individualHooks: true` is passed.

That would sum up to something like (pseudo-code):
```
runHook('beforeBulkDestroy')
if (individualHooks) {
  for each instance returned by findAll()
    instance.destroy()
} else {
  doTheBulkOperation
}
runHook('afterBulkDestroy')
```

Calling `instance.destroy()` will indeed run the before/afterDestroy hooks on it, which is what is expected by `individualHooks: true` anyway.

WDYT?

**Comment by sushantdhiman** [comment]
@ddolcimascolo Good question and your pseudo-code is exactly what we should do (like the update case), I reckon we need this change for `restore` process as well.

**Comment by ddolcimascolo** [comment]
Hey guys (and @sushantdhiman),

Can you share the status of this contribution? Do you need more code from me or can we consider it good to merge, eventually opening another issue to track the refactoring of the bulk 'destroy' operation? 

For what it's worth, we're using this patch in production for several weeks now without any issue. It fits well with https://github.com/bonaval/sequelize-temporal/pull/7 which I contributed too. 

Thanks, 
Regards, 

David

**Pull request merged**